src/actions/ProgressAction.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+declare(strict_types=1);
+
+namespace hipanel\actions;
+
+use Closure;
+use Exception;
+use Yii;
+
+class ProgressAction extends Action
+{
+    public ?Closure $onProgress = null;
+    public ?Closure $onGettingId = null;
+    private array $tries = [];
+
+    public function run()
+    {
+        $charset = Yii::$app->charset;
+        header("Content-Type: text/event-stream; charset=$charset");
+        header('Cache-Control: no-store');
+        header('Connection: keep-alive');
+        header('X-Accel-Buffering: no');
+        $id = $this->onGettingId ? call_user_func($this->onGettingId, $this) : Yii::$app->user->id;
+        $triesLeft = 250;
+        do {
+            $triesLeft--;
+            try {
+                $data = call_user_func($this->onProgress, $this);
+            } catch (Exception $exception) {
+                Yii::error(implode(PHP_EOL, [__CLASS__, $exception->getMessage()]));
+                break;
+            }
+            $this->printMessage($id, $data);
+            @ob_flush();
+            flush();
+            if (connection_aborted()) {
+                break;
+            }
+            sleep(1);
+        } while (!$this->isLimitHasBeenReached($id, $data, $triesLeft));
+        $response = Yii::$app->response;
+        $response->statusCode = 402;
+        $response->send();
+    }
+
+    private function isLimitHasBeenReached(mixed $id, string $data, int $triesLeft): bool
+    {
+        $hash = hash('sha256', implode("", [$id, $data]));
+        if (isset($this->tries[$hash]) && $this->tries[$hash] === 1) {
```

Comment:
There's a potential issue here. We've discussed that it should check whether all the last 250 messages are the same and if so, terminate the loop. However, the decrementing `$triesLeft` variable has a problem — it can become negative. This happens because on each loop iteration, we decrement `$triesLeft` from 250 without checking if it is at zero.

The condition inside `isLimitHasBeenReached()` checks if `$triesLeft` is equal to 1. If it is, it should return true and stop the loop. However, once `$triesLeft` becomes 0, it will start decrementing further into negative values, which never can be equal to 1. Since `$triesLeft` and the check inside `isLimitHasBeenReached()` doesn't account for values less than 1, the loop could run indefinitely.

You also don't need to store all the hashes. If the last hash is the same as the previous – start decrementing. Once it gets changed – reset the counter to 250.